drivers/ipm/ipm_adsp.c

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (c) 2019, Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <string.h>
+#include <device.h>
+#include <soc.h>
+#include <drivers/ipm.h>
+
+#include <platform/mailbox.h>
+#include <platform/shim.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(ipm_adsp, CONFIG_IPM_LOG_LEVEL);
+
+/* FIXME define in DTS */
+#define DT_IPM_ADSP_0_BASE_ADDRESS	0x10
+#define DT_IPM_ADSP_0_NAME		"IPM_0"
+#define DT_IPM_ADSP_0_IRQ		SOC_AGGREGATE_IRQ(7, 6)
+#define DT_IPM_ADSP_0_IRQ_PRI		3
+
+/* FIXME: set to 64 for now */
+#define IPM_ADSP_MAX_DATA_SIZE		256
+#define IPM_ADSP_MAX_ID_VAL		IPC_DIPCI_MSG_MASK
+
+/* Mailbox ADSP -> Host */
+#define IPM_ADSP_MAILBOX_OUT		MAILBOX_DSPBOX_BASE
+#define IPM_ADSP_MAILBOX_OUT_SIZE	MAILBOX_DSPBOX_SIZE
+
+/* Mailbox Host -> ADSP */
+#define IPM_ADSP_MAILBOX_IN		MAILBOX_HOSTBOX_BASE
+#define IPM_ADSP_MAILBOX_IN_SIZE	MAILBOX_HOSTBOX_SIZE
+
+struct ipm_adsp_config {
+	u8_t *base;
+	void (*irq_config_func)(struct device *dev);
+};
+
+struct ipm_adsp_data {
+	ipm_callback_t callback;
+	void *callback_ctx;
+};
+
+static void ipm_adsp_isr(void *arg)
+{
+	struct device *dev = (struct device *)arg;
+	const struct ipm_adsp_data *data = dev->driver_data;
+	uint32_t dipcctl, dipcie, dipct;
+
+	dipct = ipc_read(IPC_DIPCT);
+	dipcie = ipc_read(IPC_DIPCIE);
+	dipcctl = ipc_read(IPC_DIPCCTL);
+
+	LOG_DBG("dipct 0x%x dipcie 0x%x dipcctl 0x%x", dipct, dipcie, dipcctl);
+
+	/* IPC Target Busy Interrupt Enable (IPCTBIE) */
+	if (dipct & IPC_DIPCT_BUSY && dipcctl & IPC_DIPCCTL_IPCTBIE) {
+		/* Mask Busy interrupt */
+		ipc_write(IPC_DIPCCTL, dipcctl & ~IPC_DIPCCTL_IPCTBIE);
+
+		if (data->callback) {
+			SOC_DCACHE_INVALIDATE((void *)IPM_ADSP_MAILBOX_IN,
+					      IPM_ADSP_MAILBOX_IN_SIZE);
+			/* Use zero copy */
+			data->callback(data->callback_ctx,
+				       dipct & IPC_DIPCI_MSG_MASK,
+				       (void *)IPM_ADSP_MAILBOX_IN);
```

Comment:
Similar comment to the send side: this supports only one message in the buffer at a time.  But this buffer is 8k, and IPM is designed to be asynchronous, so some queing ability would definitely be possible and a good idea.